### PR TITLE
Add ScopeGuard implementation

### PR DESCRIPTION
**Pull request by rnestler**
- [x] Cleanup commit messages
 - [x] Add ASCII art according to the project style
 - [x] Add some documentation
 - [x] ScopeGuardList should call cleanup actions in reverse order

Inspired by
https://channel9.msdn.com/Shows/Going+Deep/C-and-Beyond-2012-Andrei-Alexandrescu-Systematic-Error-Handling-in-C

### CODE REVIEW

File: libs/librepcbcommon/scopeguard.h [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+#ifndef LIBREPCB_SCOPEGUARD_H
+#define LIBREPCB_SCOPEGUARD_H
+
+#include <utility>
+
+namespace librepcb {
+
+template<class Fun>
+class ScopeGuard
```
Comment (ubruhin) [inline libs/librepcbcommon/scopeguard.h]: If no method is declared virtual (not even the destructor), please use the "final" keyword to the class definition

File: libs/librepcbcommon/scopeguard.h [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+        ScopeGuard(Fun f) :
+            mF(std::move(f)),
+            mActive(true)
+        { }
+
+        ~ScopeGuard()
+        {
+            if (mActive) {
+                mF();
```
Comment (ubruhin) [inline libs/librepcbcommon/scopeguard.h]: if mF() throws an exception, we would have a big problem... Maybe we should add a try/catch block, and in the catch block we simple call qFatal("ScopeGuard: Internal Fatal Error") which will let the application crash ;)

**Comment by ubruhin** [comment]
Block comment would be nice :smile: 
```c++
/*****************************************************************************************
 *  Foo
 ****************************************************************************************/
```

File: libs/librepcbcommon/scopeguardlist.h [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            mScopeGuards(std::move(rhs.mScopeGuards))
+        { }
+
+        ScopeGuardList(const ScopeGuardList&) = delete;
+        ScopeGuardList& operator=(const ScopeGuardList&) = delete;
+
+        ~ScopeGuardList()
+        {
+            if (mActive) {
+                for (auto& scopeGuard : mScopeGuards) {
```
Comment (rnestler) [inline libs/librepcbcommon/scopeguardlist.h]: I think the handlers should be called in reverse order, to match the behavior of multiple local `ScopeGuard` variables.

**Comment by ubruhin** [comment]
Btw, you have referenced the wrong issue in the commit message ;) #55 would be the right one...

**Comment by ubruhin** [comment]
Great, thanks! :+1:

**Comment by ubruhin** [comment]
Oooh I've found a bug! When using the ```ScopeGuardList(size_t)``` constructor, the vector gets filled with default constructed ```std::function<void()>``` functions instead of only reserving the requested memory. This will throw a ```bad_function_call``` exception when calling such a default constructed function.

@rnestler Could you fix that? 

And in addition, it would be nice to output the catched exception in the destructor of ```ScopeGuardList``` (and also in the destructor of ```ScopeGuard```):
```C++
try{ (*scopeGuard)(); } catch(std::exception& e) {
    qFatal("Cleanup function threw an exception: %s", e.what());
}
```

**Comment by rnestler** [comment]
> the vector gets filled with default constructed std::function<void()> functions

Ohh sorry my bad. Of course. I only need to call `reserve()`

**Comment by rnestler** [comment]
Hmm btw. qAbort isn't so nice in the unit tests: 
```
[==========] Running 101 tests from 3 test cases.
[----------] Global test environment set-up.
[----------] 9 tests from ScopeGuardTest
[ RUN      ] ScopeGuardTest.testScopeGuard
[       OK ] ScopeGuardTest.testScopeGuard (0 ms)
[ RUN      ] ScopeGuardTest.testScopeGuardDismiss
[       OK ] ScopeGuardTest.testScopeGuardDismiss (0 ms)
[ RUN      ] ScopeGuardTest.testScopeGuardPerformance
Needed 0.0179434s for 10000000 loops
[       OK ] ScopeGuardTest.testScopeGuardPerformance (18 ms)
[ RUN      ] ScopeGuardTest.testScopeGuardList
[       OK ] ScopeGuardTest.testScopeGuardList (0 ms)
[ RUN      ] ScopeGuardTest.testScopeGuardListDismiss
[       OK ] ScopeGuardTest.testScopeGuardListDismiss (0 ms)
[ RUN      ] ScopeGuardTest.testScopeGuardListOrder
[       OK ] ScopeGuardTest.testScopeGuardListOrder (0 ms)
[ RUN      ] ScopeGuardTest.testScopeGuardListSizedConstructor

Program received signal SIGABRT, Aborted.
0x00007ffff6adc2a8 in raise () from /usr/lib/libc.so.6
(gdb) c
Continuing.

Program terminated with signal SIGABRT, Aborted.
The program no longer exists.
```

**Comment by ubruhin** [comment]
> Is every exception used in LibrePCB derived from std::exception?

Yes, they are. Oh now I have found a bug in the implementation of the method ```what()```, I will fix that later ;)

> Hmm btw. qAbort isn't so nice in the unit tests:

:smile: This method is handled in [libs/librepcbcommon/debug.cpp](libs/librepcbcommon/debug.cpp) but we could change its behaviour when running the unit tests...

**Comment by rnestler** [comment]
Btw. I opened #62 and hopefully fixed it in #61.

**Pull request merged**